### PR TITLE
Add endpoint to refresh auth token

### PR DESCRIPTION
**Pull request by Rushi1109**
**Changes**
- Add a new endpoint to hit, when the auth token is expired.
- Add logic in controller to refresh the auth token.

Completes the backend part for #915

### CODE REVIEW

File: Server/controllers/authController.js [hunk 1]
```diff
@@ -173,6 +173,68 @@ const loginUser = async (req, res, next) => {
...
+  try {
+    // check for refreshToken 
+    const refreshToken = req.headers["x-refresh-token"];
+
+    if (!refreshToken) {
+      // No refresh token provided
+      const error = new Error(errorMessages.NO_REFRESH_TOKEN);
+      error.status = 401;
+      error.service = SERVICE_NAME;
+      error.method = "handleExpiredJwtToken";
```
Comment (marcelluscaio) [inline Server/controllers/authController.js:196]: Here we do not have a refreshToken. should the method be handleExpiredJwtToken?

Comment (ajhollid) [reply Server/controllers/authController.js:196]: `error.method` is meant to refer to the method that is throwing the error so we have more information in the logs to quickly track down where errors are being thrown.

In this case the method name should be `refreshAuthToken`

**Review by ajhollid — CHANGES_REQUESTED** [review]
Looks pretty good to me for the most part :+1: 

The `error.method` value should be the name of the method that is throwing the error, so that ca be updated.

Please include updated tests for `authController` and document the new endpoint in the `openapi.json` file so we have complete documentation and tests.

File: Server/controllers/authController.js (hunk 1 above)
Comment (Rushi1109) [reply Server/controllers/authController.js:196]: > `error.method` is meant to refer to the method that is throwing the error so we have more information in the logs to quickly track down where errors are being thrown.
> 
> In this case the method name should be `refreshAuthToken`

Yeah. I am aware of that, I copied the code from the middleware, so forgot to update the method name there. I'll update it.

**Comment by Rushi1109** [comment]
> Please include updated tests for `authController` and document the new endpoint in the `openapi.json` file so we have complete documentation and tests.

Hello @ajhollid,
I tried adding test cases but some lines are uncovered. I am having difficulties, so I would need some help here.

**Comment by ajhollid** [comment]
> > Please include updated tests for `authController` and document the new endpoint in the `openapi.json` file so we have complete documentation and tests.
> 
> Hello @ajhollid, I tried adding test cases but some lines are uncovered. I am having difficulties, so I would need some help here.

Hi @Rushi1109,

I'd be happy to help with some test cases, but I would need you to grant me access to your forked repo so I can push to this branch.

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by Rushi1109** [comment]
> I'd be happy to help with some test cases, but I would need you to grant me access to your forked repo so I can push to this branch.

Hello @ajhollid,

image
I guess It was enabled by default.

**Comment by ajhollid** [comment]
> > I'd be happy to help with some test cases, but I would need you to grant me access to your forked repo so I can push to this branch.
> 
> Hello @ajhollid,
> 
> image
> I guess It was enabled by default.

OK ill give it a try when I have a chance. It's the weekend coming up so it might be a couple days. I will get to it as soon as I can 👍

**Comment by Rushi1109** [comment]
> OK ill give it a try when I have a chance. It's the weekend coming up so it might be a couple days. I will get to it as soon as I can 👍

Sure.

**Comment by ajhollid** [comment]
Hey @Rushi1109 ,

I refactored the function slightly, I'm not sure why all code following verifying refresh token was carried out in the callback, but I don't think there's any reason to. Please review the changes I made and make sure everything is still functioning as you expect.

I also added the missing test cases so you've got 100% coverage again now.

**Comment by ajhollid** [comment]
@Rushi1109 do you still want to finish this PR?  I will close it otherwise.

Thanks!

**Comment by Rushi1109** [comment]
Hello @ajhollid, 
Sorry for the inactivity.

> I also added the missing test cases so you've got 100% coverage again now.
Thanks for adding those test cases.

> I refactored the function slightly, I'm not sure why all code following verifying refresh token was carried out in the callback, but I don't think there's any reason to. Please review the changes I made and make sure everything is still functioning as you expect.
Yeah I guess it should not be inside the callback. Thanks for correction.

I have reviewed the changes and it looks good to me.

**Comment by Rushi1109** [comment]
I think the PR is ready to be merged now.

@ajhollid.

**Comment by ajhollid** [comment]
> Hello @ajhollid, 
> Sorry for the inactivity.
> 
> > I also added the missing test cases so you've got 100% coverage again now.
> 
> Thanks for adding those test cases.
> 
> > I refactored the function slightly, I'm not sure why all code following verifying refresh token was carried out in the callback, but I don't think there's any reason to. Please review the changes I made and make sure everything is still functioning as you expect.
> 
> Yeah I guess it should not be inside the callback. Thanks for correction.
> 
> I have reviewed the changes and it looks good to me.

No problem, just wanted to make sure you were still with us 😂

**Comment by ajhollid** [comment]
> I think the PR is ready to be merged now.
> 
> @ajhollid.

Awesome thanks for your work! I will review as soon as I can 👍

**Comment by Rushi1109** [comment]
> No problem, just wanted to make sure you were still with us 😂

Yeah, It was a weekend. So, I wasn't working on it.

**Review by ajhollid — APPROVED** [review]
Looks good to me, thanks for making revisions :+1:

**Pull request merged**